### PR TITLE
json and bank stuff

### PR DESCRIPTION
**Pull request by lun-4**
- json: atomic write thing, why not
 - bank: add `Bank.transfer`

### CODE REVIEW

File: kmn/cogs/currency.py [hunk 1]
```diff
@@ -31,13 +31,22 @@ def get_account(self, user):
         account = self.storage.get(str(user.id))
 
         if not account:
-            raise CommandFailure('you have no account, create one with `{prefix}register`.')
+            raise CommandFailure('you have no account, create one '
```
Comment (slice) [inline kmn/cogs/currency.py:34]: revert: i use 120 character long lines

File: kmn/cogs/currency.py [hunk 2]
```diff
@@ -49,15 +58,16 @@ def __init__(self, bot):
...
+        balance = self.bank.wallet(who)
         await ctx.send(f'{who} \N{RIGHTWARDS ARROW} `{balance:.2f}bc`')
 
     @command()
     @is_bot_admin()
     async def write(self, ctx, user: discord.User, amount: float):
         """write someone's wallet lol"""
         await self.bank.write(user, amount)
-        await ctx.send(f'\N{OK HAND SIGN} {user} \N{RIGHTWARDS ARROW} `{amount:.2f}bc`')
+        await ctx.send(f'\N{OK HAND SIGN} {user} \N{RIGHTWARDS ARROW}'
```
Comment (slice) [inline kmn/cogs/currency.py:69]: ditto

File: kmn/cogs/currency.py [hunk 3]
```diff
@@ -70,41 +80,44 @@ def __init__(self, bot):
...
         # grab thief wallet
-        thief_wallet = await self.bank.wallet(ctx.author)
+        thief_wallet = self.bank.wallet(ctx.author)
 
         # check wallet
-        target_wallet = await self.bank.wallet(target)
+        target_wallet = self.bank.wallet(target)
         if amount > target_wallet:
-            raise CommandFailure("that person doesn't have the money you want. look for someone else.")
+            raise CommandFailure("that person doesn't have the "
```
Comment (slice) [inline kmn/cogs/currency.py:88]: ditto

File: kmn/cogs/currency.py [hunk 4]
```diff
@@ -70,41 +80,44 @@ def __init__(self, bot):
...
+                                 "money you want. look for someone else.")
 
         # can't steal more than 50%
         percentage_of_steal = amount / target_wallet
         if percentage_of_steal > 0.5:
             amount = target_wallet * 0.5
             raise CommandFailure(
-                f"you can't steal more than 50% of someone's wallet. the max you can steal from {target} is"
+                f"you can't steal more than 50% of someone's wallet. "
```
Comment (slice) [inline kmn/cogs/currency.py:96]: ditto

File: kmn/cogs/currency.py [hunk 5]
```diff
@@ -70,41 +80,44 @@ def __init__(self, bot):
...
-            await self.bank.write(ctx.author, thief_wallet + amount)
+            await self.bank.transfer(target, ctx.author, amount)
             await ctx.send(
                 f'\N{PISTOL} ouch. {ctx.author} \N{RIGHTWARDS ARROW} `{amount:.2f}bc`'
                 f'\n\n[amount:`{amount}` < rolled:`{chance:.2f}`] chance = `{potential:.2f}`'
             )
         else:
             if thief_wallet == 0:
-                return await ctx.send('\N{ONCOMING POLICE CAR} oops, you failed.')
+                return await ctx.send('\N{ONCOMING POLICE CAR} '
```
Comment (slice) [inline kmn/cogs/currency.py:112]: ditto

File: kmn/cogs/currency.py [hunk 6]
```diff
@@ -70,41 +80,44 @@ def __init__(self, bot):
...
+                                      'oops, you failed.')
 
             penalty = min(amount * 1.5, thief_wallet)
             percentage = penalty / thief_wallet
             await self.bank.write(ctx.author, thief_wallet - penalty)
             await ctx.send(
-                f'\N{ONCOMING POLICE CAR} oops, you failed. you have lost `{penalty:.2f}bc`'
+                f'\N{ONCOMING POLICE CAR} oops, you failed. '
```
Comment (slice) [inline kmn/cogs/currency.py:119]: ditto

**Comment by lun-4** [comment]
note: untested code

File: kmn/cogs/currency.py [hunk 7]
```diff
@@ -115,30 +128,29 @@ def __init__(self, bot):
...
 
-        # deduct from balance
-        await self.bank.write(ctx.author, wallet - amount)
-
-        # write!
-        await self.bank.write(target, target_wallet + amount)
+        await self.bank.transfer(ctx.author, target, amount)
 
-        await ctx.send(f'\N{MONEY WITH WINGS} {ctx.author} `--[ {amount:.2f}bc ]-->` {target}')
+        await ctx.send(f'\N{MONEY WITH WINGS} {ctx.author} '
```
Comment (slice) [inline kmn/cogs/currency.py:147]: ditto

File: kmn/cogs/currency.py [hunk 8]
```diff
@@ -115,30 +128,29 @@ def __init__(self, bot):
...
+                       f'`--[ {amount:.2f}bc ]-->` {target}')
 
         try:
             await target.send(
-                f'\N{MONEY WITH WINGS} you got money! you have received `{amount:.2f}bc` from {ctx.author.mention}.'
+                f'\N{MONEY WITH WINGS} you got money! you have '
```
Comment (slice) [inline kmn/cogs/currency.py:152]: ditto

File: kmn/storage.py [hunk 9]
```diff
@@ -15,9 +16,13 @@ def __init__(self, file_name: str, *, loop):
             self._data = {}
 
     def _save(self):
-        with open(self.file_name, 'w') as fp:
+        filename = self.file_name + '.save'
+        with open(filename, 'w') as fp:
             json.dump(self._data, fp)
 
+        os.rename(filename, self.file_name)
+        os.remove(filename)
```
Comment (slice) [inline kmn/storage.py:24]: use `os.replace`

**Review by slice — CHANGES_REQUESTED** [review]
lol

**Pull request closed without merging**